pytorch_lightning/core/optimizer.py

```diff
@@ -0,0 +1,250 @@
+
+                def gen_closure():
+                    ... forward and compute loss for generator
+                    loss_gen = ...
+                    self.log("loss_gen", loss_gen, on_step=True, on_epoch=True)
+                    self.manual_backward(loss_gen, opt_gen)
+
+                def dis_closure():
+                    ... forward and compute loss for discriminator
+                    loss_dis = ...
+                    self.log("loss_dis", loss_dis, on_step=True, on_epoch=True)
+                    self.manual_backward(loss_dis, opt_dis)
+
+                # this will accumulate gradients for 2 batches and then call opt_gen.step()
+                opt_gen.step(closure=gen_closure, make_optimizer_step=batch_idx % 2 == 0)
+
+                # update discriminator every 4 batches
+                # therefore, no gradient accumulation for discriminator
+                if batch_idx % 4 == 0 :
+                    # Note: Set make_optimizer_step to True or it will use by default
+                    # Trainer(accumulate_grad_batches=x)
+                    opt_dis.step(closure=optimizer_closure, make_optimizer_step=True)
+        """
+
+        profiler_name = "optimizer_step_and_closure"
+        if closure is None:
+            closure = do_nothing_closure
+            profile_name = "optimizer_step"
+        else:
+            if not isinstance(closure, types.FunctionType):
+                raise MisconfigurationException("When closure is provided, it should be a function")
+
+        if make_optimizer_step is None:
+            make_optimizer_step = not self._should_accumulate
+
+        trainer = self._trainer
+        optimizer = self._optimizer
+
+        if make_optimizer_step:
+            if trainer.on_tpu:
+                with trainer.profiler.profile(profiler_name):
+                    xm.optimizer_step(optimizer, optimizer_args={'closure': closure, **kwargs})
+
+            elif trainer.amp_backend is not None:
+                trainer.amp_backend.step(trainer, optimizer, closure)
+
+            else:
+                with trainer.profiler.profile(profiler_name):
+                    optimizer.step(closure=closure, *args, **kwargs)
+
+            # perform zero grad
+            optimizer.zero_grad()
+        else:
+            # make sure to call optimizer_closure when accumulating
+            with trainer.train_loop.block_ddp_sync_behaviour():
+                with trainer.profiler.profile("closure"):
+                    closure()
+
+    def __repr__(self):
+        format_string = "Lightning" + self._optimizer.__class__.__name__ + ' ('
+        for i, group in enumerate(self.param_groups):
+            format_string += '\n'
+            format_string += 'Parameter Group {0}\n'.format(i)
+            for key in sorted(group.keys()):
+                if key != 'params':
+                    format_string += '    {0}: {1}\n'.format(key, group[key])
+        format_string += ')'
+        return format_string
```

Comment:
I think this representation is more suited for `__str__` considering it would be hard to parse with all the extra whitespace.

quoting https://stackoverflow.com/a/2626364/7884046:

> __repr__ goal is to be unambiguous
> __str__ goal is to be readable

Reply:
An idea for a better repr is:

```python
LightningOptimizer(name='SGD', groups=[{"learning_rate": 0.01, "momentum": 0, ...}, ...])
```

Reply:
```
LightningOptimizer(optim=SGD,groups=[(dampening=0,initial_lr=0.1,lr=0.01,momentum=0,nesterov=0,weight_decay=0)])
```

Reply:
I would keep the quotes for the values, also the dict for the group items, considering that is their actual structure internally.

Have a look at how python would generate the `__repr__`
```
>>> @dataclass
... class Foo:
...     a: str
...     b: dict
... 
>>> Foo('SGD', dict(lr=0.01))
Foo(a='SGD', b={'lr': 0.01})
```